torch/distributed/optim/named_optimizer.py

```diff
@@ -0,0 +1,244 @@
+import logging
+import warnings
+
+from copy import deepcopy
+from typing import Any, Collection, Dict, List, Mapping, Union
+
+import torch
+from torch import optim
+from torch.distributed._shard.sharded_tensor import ShardedTensor
+
+
+__all__ = ["NamedOptimizer"]
+
+logger = logging.getLogger(__name__)
+
+
+class NamedOptimizer(optim.Optimizer):
+    """
+    NamedOptimizer takes a dict of parameters and exposes state_dict by parameter key.
```

Comment:
nit: As a heads up, if we want this class to be public in the future, we need to use double backtick (\'\') to surround text that should be rendered as code (e.g. \'\'state_dict\'\' here).

Reply:
Ah I see. Thanks!